canonical/istio-k8s-operator
# Add dashboards integration to istio charms

## Issue
This PR adds necessary grafana dashboards to integrate with COS


## Solution
Dashboards are based on upstream [istio dashboards](https://istio.io/latest/docs/ops/integrations/grafana/#configuration) with a lot of modifications to make it work with ambient, some of these modifications are updating prometheus source to match cos and reporter type to be `waypoint`


## Context
[Mesh Dashboard](https://grafana.com/grafana/dashboards/7639) provides an overview of all services in the mesh.
[Service Dashboard](https://grafana.com/grafana/dashboards/7636) provides a detailed breakdown of metrics for a service.
[Workload Dashboard](https://grafana.com/grafana/dashboards/7630) provides a detailed breakdown of metrics for a workload.
[Performance Dashboard](https://grafana.com/grafana/dashboards/11829) monitors the resource usage of the mesh.
[Control Plane Dashboard](https://grafana.com/grafana/dashboards/7645) monitors the health and performance of the control plane.
[Ztunnel Dashboard](https://grafana.com/grafana/dashboards/21306-istio-ztunnel-dashboard/) monitors the health and performance of ztunnel.


## Testing Instructions
Testing can be done in many ways, but mainly it should be

1. Getting metrics from istio-k8s and istio-beacon-k8s into cos
2. Putting workloads on the mesh
3. Relating istio-k8s to grafana to get dashboards
4. Generate some traffic inside the mesh
5. Explore the dashboards

Below are step by step instructions on how to do this
```bash
juju add-model cos
curl -L https://raw.githubusercontent.com/canonical/cos-lite-bundle/main/overlays/offers-overlay.yaml -O
juju deploy cos-lite -m cos --channel edge --trust --overlay ./offers-overlay.yaml

charmcraft pack
juju add-model istio-system
juju deploy ./istio-k8s_ubuntu-22.04-amd64.charm -m istio-system --trust --resource metrics-proxy-image=docker.io/ibraaoad/metrics-proxy:v0.1.0
juju deploy grafana-agent-k8s --channel edge --trust
juju consume admin/cos.grafana-dashboards
juju consume admin/cos.prometheus-receive-remote-write
juju relate grafana-agent-k8s:send-remote-write prometheus-receive-remote-write
juju relate grafana-agent-k8s:grafana-dashboards-provider grafana-dashboards
juju relate grafana-agent-k8s:grafana-dashboards-consumer istio-k8s:grafana-dashboard
juju relate grafana-agent-k8s:metrics-endpoint istio-k8s:metrics-endpoint


juju add-model mesh-model
juju deploy loki-k8s --trust --channel edge
juju deploy flog-k8s --trust --channel edge #to generate traffic in the mesh
juju deploy grafana-agent-k8s --channel edge --trust
juju deploy istio-beacon-k8s --channel edge --trust --config model-on-mesh=true
juju consume admin/cos.prometheus-receive-remote-write
juju relate flog-k8s loki-k8s
juju relate grafana-agent-k8s:send-remote-write prometheus-receive-remote-write
juju relate grafana-agent-k8s:metrics-endpoint istio-beacon-k8s:metrics-endpoint

juju run -m cos grafana/0 get-admin-password

#Explore dashboards :D
```
